### PR TITLE
boards: stm32f746g_disco ft5336 support, plus sample application

### PR DESCRIPTION
**Pull request by k0d**
The first commit adds support for the ft5336 touch panel on the stm32f746g_disco board. It enables the driver and adds the relevant device tree config. It's dependent on the I2C_3 device so this is enabled also.

The second commit adds a sample application so it can be tested. It's tested on the stm32f746g_disco board, but should work on other boards for example the mimxrt1060_evk.

### CODE REVIEW

**Comment by k0d** [comment]
@MaureenHelm Could you test the sample on the mimxrt1060_evk board please?

'zephyr/samples/drivers/kscan_ft5336'

File: samples/drivers/kscan_ft5336/sample.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+tests:
+  sample.drivers.kscan_ft5336:
+    tags: drivers
+    harness: console
+    harness_config:
+        type: multi_line
+        ordered: true
+        regex:
+            - "kb data(.*)"
+    depends_on: kscan
```
Comment (MaureenHelm) [inline samples/drivers/kscan_ft5336/sample.yaml]: You'll want to add `kscan` to the supported list in `boards/arm/stm32f746g_disco/stm32f746g_disco.yaml` to ensure this sample builds for this board in CI. Do it in commit `boards: Configure ft5336 touch panel on stm32f746g_disco board`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> @MaureenHelm Could you test the sample on the mimxrt1060_evk board please?

I tested on `mimxrt1050_evk` and it worked after I made a small change:
```diff
diff --git a/boards/arm/mimxrt1050_evk/Kconfig.defconfig b/boards/arm/mimxrt1050_evk/Kconfig.defconfig
index 9763200fef..c338f71ac5 100644
--- a/boards/arm/mimxrt1050_evk/Kconfig.defconfig
+++ b/boards/arm/mimxrt1050_evk/Kconfig.defconfig
@@ -40,7 +40,7 @@ config GPIO_MCUX_IGPIO_5
 endif # GPIO_MCUX_IGPIO

 config I2C
-       default y if LVGL
+       default y if KSCAN

 config I2C_1
        default y
```

File: samples/drivers/kscan_ft5336/sample.yaml (hunk 1 above)
Comment (k0d) [reply samples/drivers/kscan_ft5336/sample.yaml]: I changed it to kscan_ft5336 so it doesn't run the other kscan sample build.

**Comment by k0d** [comment]
> > @MaureenHelm Could you test the sample on the mimxrt1060_evk board please?
> 
> I tested on `mimxrt1050_evk` and it worked after I made a small change:
> 
I've added it in a separate commit and enabled the kscan_ft5336 sample for the 3 imx boards that have the ft5336 chip. I can't test it, but it's such a simple change it will work. I should get one of those boards at some point...

**Review by MaureenHelm — APPROVED** [review]
> I've added it in a separate commit and enabled the kscan_ft5336 sample for the 3 imx boards that have the ft5336 chip. I can't test it, but it's such a simple change it will work. I should get one of those boards at some point...

Thank you! I tested the updated PR again on `mimxrt1050_evk` and it works.

File: samples/drivers/kscan_ft5336/src/main.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2020 Mark Olsson <mark@markolsson.se>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
```
Comment (jfischer-phytec-iot) [inline samples/drivers/kscan_ft5336/src/main.c]: Please make it generic, the sample should be for the driver API not for specific driver.

**Review by jfischer-no — DISMISSED** [review]

File: samples/drivers/kscan_ft5336/src/main.c (hunk 2 above)
Comment (k0d) [reply samples/drivers/kscan_ft5336/src/main.c]: Isn't the other sample specific to another driver? I guess it's better to change it to be a touch sample?

Comment (jfischer-phytec-iot) [reply samples/drivers/kscan_ft5336/src/main.c]: > Isn't the other sample specific to another driver?

Indeed. Could you rename your sample to something like kscan_touch and make it more generic?

File: boards/arm/mimxrt1050_evk/Kconfig.defconfig [hunk 3]
```diff
@@ -40,7 +40,7 @@ config GPIO_MCUX_IGPIO_5
 endif # GPIO_MCUX_IGPIO
 
 config I2C
-	default y if LVGL
+	default y if KSCAN
```
Comment (erwango) [inline boards/arm/mimxrt1050_evk/Kconfig.defconfig]: What about `if (LVGL || KSCAN)` ?
We might want to enable LVGL w/o KSCAN. This is not what is done below in this file, but this is questionable.
@MaureenHelm, what is your opinion ?

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts [hunk 4]
```diff
@@ -18,6 +18,10 @@
 		zephyr,flash = &flash0;
 		zephyr,dtcm = &dtcm;
 	};
+	
+	aliases {
+		kscan0 = &touch_controller;
```
Comment (erwango) [inline boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: Please move this to the existing aliases section below.

File: boards/arm/stm32f746g_disco/Kconfig.defconfig [hunk 5]
```diff
@@ -38,4 +45,17 @@ config PWM_STM32_1
 	default y
 	depends on PWM
 
+config KSCAN
+	default y if LVGL
```
Comment (erwango) [inline boards/arm/stm32f746g_disco/Kconfig.defconfig]: I don't think we should enable KSCAN depending on LVGL (unless there is a dependency in the drivers itself, but ideally it should not be the case.).

**Review by erwango — CHANGES_REQUESTED** [review]
Looks fine. Few changes requested.
Main concern is the dependency w/ LVGL.

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts (hunk 4 above)
Comment (k0d) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: Oops...missed that as it was lower down.

File: boards/arm/stm32f746g_disco/Kconfig.defconfig (hunk 5 above)
Comment (k0d) [reply boards/arm/stm32f746g_disco/Kconfig.defconfig]: ok, I can agree with that.

File: boards/arm/mimxrt1050_evk/Kconfig.defconfig (hunk 3 above)
Comment (k0d) [reply boards/arm/mimxrt1050_evk/Kconfig.defconfig]: I'll do whatever you think is best on this one. Lower down in the file is...
```config KSCAN
	default y if LVGL
```
So there is some dependancy chain there.

Comment (MaureenHelm) [reply boards/arm/mimxrt1050_evk/Kconfig.defconfig]: The dependency is for KSCAN; LVGL doesn't use I2C.

Comment (k0d) [reply boards/arm/mimxrt1050_evk/Kconfig.defconfig]: So the question is if we have KSCAN as a dependancy of LVGL?

**Review by erwango — APPROVED** [review]
I'm approving the change.

Remaining point is the fact that KSCAN depends on LVGL activation in imx board configuration files, but:
- this is not directly related to this change
- it's up to @MaureenHelm to give the final word on these boards.

**Review by MaureenHelm — APPROVED** [review]

**Comment by k0d** [comment]
I will change this to use the new DT_INST macros that galek has been working on...then I think this is ready to be merged.

**Comment by k0d** [comment]
I've changed this to use the new DT_INST macros that @galak  has been working on...

I'm happy with the changes now and I think this is ready to be merged.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@k0d ; can you fix shippable ?

**Comment by MaureenHelm** [comment]
> @k0d ; can you fix shippable ?

@k0d ping

**Comment by erwango** [comment]
@kod, can you rebase ?

**Comment by k0d** [comment]
I've rebased...will see if shippable works...otherwise I'll address this later today. I've been working on a different project the past 2 weeks...so haven't had time to look at this until now.

File: boards/arm/stm32f746g_disco/Kconfig.defconfig [hunk 6]
```diff
@@ -26,6 +26,16 @@ config UART_6
...
+config I2C
+	default y if KSCAN
+
+config I2C_1
+	default y
+	depends on I2C
+
+config I2C_3
+	default y if KSCAN
+	depends on I2C
```
Comment (erwango) [inline boards/arm/stm32f746g_disco/Kconfig.defconfig:38]: Please remove these. cf https://github.com/zephyrproject-rtos/zephyr/pull/24342

File: boards/arm/stm32f746g_disco/pinmux.c [hunk 7]
```diff
@@ -42,6 +42,10 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB8, STM32F7_PINMUX_FUNC_PB8_I2C1_SCL},
 	{STM32_PIN_PB9, STM32F7_PINMUX_FUNC_PB9_I2C1_SDA},
 #endif  /* CONFIG_I2C_1 */
+#ifdef CONFIG_I2C_3
```
Comment (erwango) [inline boards/arm/stm32f746g_disco/pinmux.c:45]: #if DT_HAS_NODE(DT_NODELABEL(i2c3))

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by k0d** [comment]
d

**Comment by k0d** [comment]
@erwango Shippable fixed, all seems to be good now. Ready to merge I think.

File: drivers/kscan/kscan_handlers.c [hunk 8]
```diff
@@ -11,21 +11,21 @@ static inline int z_vrfy_kscan_config(struct device *dev,
...
+static inline int z_vrfy_kscan_disable_callback(struct device *dev)
 {
 	Z_OOPS(Z_SYSCALL_DRIVER_KSCAN(dev, disable_callback));
 
 	return z_impl_kscan_disable_callback((struct device *)dev);
 }
 #include <syscalls/kscan_disable_callback_mrsh.c>
 
-static int z_vrfy_kscan_enable_callback(struct device *dev);
+static int z_vrfy_kscan_enable_callback(struct device *dev)
```
Comment (MaureenHelm) [inline drivers/kscan/kscan_handlers.c:28]: The changes in this file don't appear to be related to the commit message, `tests: drivers: kscan fixes failing test due to wrong devicetree define`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/kscan/kscan_handlers.c (hunk 8 above)
Comment (k0d) [reply drivers/kscan/kscan_handlers.c:28]: Fixed!

**Review by MaureenHelm — APPROVED** [review]
@erwango are you good with this?

**Review by erwango — APPROVED** [review]

**Comment by MaureenHelm** [comment]
Closing and reopening to retrigger CI checks that never returned

**Pull request merged**